cache/refs.go

```
@@ -631,109 +628,116 @@ func (sr *immutableRef) ociDesc(ctx context.Context, dhs DescHandlers) (ocispecs
+}
 
-	info, err = cs.Info(ctx, desc.Digest)
+func (sr *immutableRef) getBlobWithCompression(ctx context.Context, compressionType compression.Type) (ocispecs.Descriptor, error) {
+	if _, err := sr.cm.ContentStore.Info(ctx, sr.getBlob()); err != nil {
+		return ocispecs.Descriptor{}, err
+	}
+	desc, err := sr.ociDesc(ctx, nil)
 	if err != nil {
-		return err
+		return ocispecs.Descriptor{}, err
 	}
-	var fields []string
-	info.Labels = map[string]string{
-		compressionVariantMediaTypeLabel: desc.MediaType,
+	return getBlobWithCompression(ctx, sr.cm.ContentStore, desc, compressionType)
+}
+
+func getBlobWithCompression(ctx context.Context, cs content.Store, desc ocispecs.Descriptor, compressionType compression.Type) (ocispecs.Descriptor, error) {
+	if compressionType == compression.UnknownCompression {
+		return ocispecs.Descriptor{}, fmt.Errorf("cannot get unknown compression type")
 	}
-	fields = append(fields, "labels."+compressionVariantMediaTypeLabel)
-	for k, v := range filterAnnotationsForSave(desc.Annotations) {
-		k2 := compressionVariantAnnotationsLabelPrefix + k
-		info.Labels[k2] = v
-		fields = append(fields, "labels."+k2)
+	var target *ocispecs.Descriptor
+	if err := walkBlob(ctx, cs, desc, func(desc ocispecs.Descriptor) bool {
+		if needs, err := needsConversion(ctx, cs, desc, compressionType); err == nil && !needs {
+			target = &desc
+			return false
+		}
+		return true
+	}); err != nil {
+		return ocispecs.Descriptor{}, err
 	}
-	if _, err := cs.Update(ctx, info, fields...); err != nil {
-		return err
+	if target == nil {
+		return ocispecs.Descriptor{}, errdefs.ErrNotFound
 	}
+	return *target, nil
+}
 
+func walkBlob(ctx context.Context, cs content.Store, desc ocispecs.Descriptor, f func(ocispecs.Descriptor) bool) error {
+	if !f(desc) {
+		return nil
+	}
+	if _, err := walkBlobVariantsOnly(ctx, cs, desc.Digest, func(desc ocispecs.Descriptor) bool { return f(desc) }, nil); err != nil {
+		logrus.WithError(err).Warn("failed to get variant blob")
```

Why not also return the error? Or if we don't want to return an error from this func, should we just drop the `error` return type?